### PR TITLE
CamelCasePlugin option to maintainJson data format

### LINKED ISSUE #281: [CamelCasePlugin] Add option to disable transforming of JSON content
When using the CamelCasePlugin, it also transforms the content of JSON fields.
It would be nice to have an option that disables that, so the contents remain in snake_case.

Discord thread for reference:
https://discord.com/channels/890118421587578920/890118421587578925/1063741659545882625

### PR DESCRIPTION
**Pull request by otaviosoares**
closes #281

### CODE REVIEW

File: test/node/src/camel-case.test.ts [hunk 1]
```diff
@@ -203,5 +209,33 @@ for (const dialect of BUILT_IN_DIALECTS) {
...
+    if (dialect !== 'sqlite') {
+      it('should respect maintainJsonFormat', async () => {
+        const dbWithOption = new Kysely<CamelDatabase>({
+          ...ctx.config,
+          plugins: [
+            new CamelCasePlugin({
+              maintainJsonFormat: true,
+            }),
+          ],
+        })
```
Comment (igalklebanov) [inline test/node/src/camel-case.test.ts:223]: let's reuse `camelDb`..

```ts
camelDb.withoutPlugins().withPlugin(new CamelCasePlugin({ maintainJsonFormat: true }))
```

File: test/node/src/camel-case.test.ts [hunk 2]
```diff
@@ -203,5 +209,33 @@ for (const dialect of BUILT_IN_DIALECTS) {
         },
       })
     })
+
+    // Can't run this test on SQLite because json field returns string
+    if (dialect !== 'sqlite') {
```
Comment (igalklebanov) [inline test/node/src/camel-case.test.ts:214]: could we cover sqlite as well? it's a `JSON.parse` or a an assertion against a string away..

File: test/node/src/camel-case.test.ts [hunk 3]
```diff
@@ -203,5 +209,33 @@ for (const dialect of BUILT_IN_DIALECTS) {
...
+              maintainJsonFormat: true,
+            }),
+          ],
+        })
+
+        const data = await dbWithOption
+          .selectFrom('camelPerson')
+          .selectAll()
+          .executeTakeFirst()
+        expect(data?.preferences).to.eql({
```
Comment (igalklebanov) [inline test/node/src/camel-case.test.ts:229]:
```suggestion
          .executeTakeFirstOrThrow()
          
        expect(data.preferences).to.eql({
```

File: test/node/src/camel-case.test.ts [hunk 4]
```diff
@@ -203,5 +209,33 @@ for (const dialect of BUILT_IN_DIALECTS) {
...
+          disable_emails: true,
+        })
+
+        await dbWithOption.destroy()
```
Comment (igalklebanov) [inline test/node/src/camel-case.test.ts:233]: no need if reusing `camelDb`.

File: src/plugin/camel-case/camel-case-plugin.ts [hunk 5]
```diff
@@ -166,8 +173,15 @@ export class CamelCasePlugin implements KyselyPlugin {
   }
 }
 
-function canMap(obj: unknown): obj is Record<string, unknown> {
+function canMap(
+  obj: unknown,
+  opt?: CamelCasePluginOptions
```
Comment (igalklebanov) [inline src/plugin/camel-case/camel-case-plugin.ts:178]:
```suggestion
  opt: CamelCasePluginOptions
```

File: src/plugin/camel-case/camel-case-plugin.ts [hunk 6]
```diff
@@ -166,8 +173,15 @@ export class CamelCasePlugin implements KyselyPlugin {
...
+): obj is Record<string, unknown> {
   return (
-    isObject(obj) && !isDate(obj) && !isBuffer(obj) && !isArrayBufferOrView(obj)
+    isObject(obj) &&
+    !isDate(obj) &&
+    !isBuffer(obj) &&
+    !isArrayBufferOrView(obj) &&
+    !opt?.maintainJsonFormat
```
Comment (igalklebanov) [inline src/plugin/camel-case/camel-case-plugin.ts:185]:
```suggestion
    !opt.maintainJsonFormat
```

**Review by igalklebanov — CHANGES_REQUESTED** [review]
Hey 👋 

Thank you! 💪

**Comment by igalklebanov** [comment]
@koskimas wondering if transformer should apply naming strategy recursively to json parameters, seems not aligned with output transformation logic.. WDYT?

File: test/node/src/camel-case.test.ts (hunk 1 above)
Comment (otaviosoares) [reply test/node/src/camel-case.test.ts]: Thank, didn't know about that one

**Review by igalklebanov — APPROVED** [review]
LGTM! 🚀

**Comment by koskimas** [comment]
`maintainJsonFormat` is not the best name for the option IMO. This doesn't just apply to JSON but any nested objects. A better name would be `maintainNestedObjectKeys`

**Comment by koskimas** [comment]
> @koskimas wondering if transformer should apply naming strategy recursively to json parameters, seems not aligned with output transformation logic.. WDYT?

You mean essentially making this option `true` by default? The use case I was thinking when I implemented this wasn't only json object columns but `json_agg` etc. nested objects that would by default have snake_cased properties you'd also want to map.

I'm fine by changing the default, but It could be a nasty breaking change.

**Comment by koskimas** [comment]
LGTM! Thank you @otaviosoares ❤️

**Comment by igalklebanov** [comment]
> > @koskimas wondering if transformer should apply naming strategy recursively to json parameters, seems not aligned with output transformation logic.. WDYT?
> 
> You mean essentially making this option `true` by default? The use case I was thinking when I implemented this wasn't only json object columns but `json_agg` etc. nested subqueries that would by default have snake_cased properties you'd also want to map.
> 
> I'm fine with changing the default, but It could be a nasty breaking change.

Sorry, I meant that we seem to only alter identifiers on input, doing nothing to json parameters in value nodes. On output, we touch everything.

Makes you wonder, if for predictability, consumers would want the option that nested json keys have same naming strategy as columns.

**Pull request merged**